src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs

```diff
@@ -371,15 +513,21 @@ private void CheckIncompleteHeaderBlock(bool endHeaders)
             }
         }
 
-        private void ProcessHeaderValue(IHttpHeadersHandler? handler)
+        private void ProcessHeaderValue(ReadOnlySpan<byte> data, IHttpHeadersHandler? handler)
         {
-            OnString(nextState: State.Ready);
+            var headerNameSpan = _headerNameRange.Equals(default)
+                ? new Span<byte>(_headerName, 0, _headerNameLength)
+                : data[_headerNameRange];
 
-            var headerNameSpan = new Span<byte>(_headerName, 0, _headerNameLength);
-            var headerValueSpan = new Span<byte>(_headerValueOctets, 0, _headerValueLength);
+            var headerValueSpan = _headerValueRange.Equals(default)
+                ? new Span<byte>(_headerValueOctets, 0, _headerValueLength)
+                : data[_headerValueRange];
 
             handler?.OnHeader(headerNameSpan, headerValueSpan);
```

Comment:
Is it commonly the case that handler is null and _index is false?  If yes, all of this computing of slices isn't required.

Reply:
I'm not sure why handler is nullable in this code. ASP.NET Core always passes in a handler. I'll check whether runtime ever passes null. This might not be needed.

Reply:
I think HttpClient passes null here to ignore headers on an intermediate response, e.g. 100 Continue. But if it's an issue, we could easily pass in a dummy handler that just ignores all the callbacks.

Reply:
It looks like it stems from here:
https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/Http2Connection.cs,365

Reply:
@geoffkizer, I like your suggestion of in this case passing in a singleton dummy interface implementation for that case; we can then eliminate all those unnecessary null checks from the code, and even if it doesn't matter for perf, generally simplify it.

Reply:
The "closed stream" case should be very rare, so hopefully it would not interfere with the interface dispatch perf.

Reply:
I won't change the decoder to assume the handler is never null in this PR - I don't want to start messing around in System.Net.Http code - but you're welcome to change it later. I don't think any changes are required in ASP.NET Core, but I'll make them if necessary.